src/provider/model/ClientHandler.ts

```diff
@@ -0,0 +1,93 @@
+import {Identity} from 'openfin/_v2/main';
+import {ApplicationOption} from 'openfin/_v2/api/application/applicationOption';
+import {injectable, inject} from 'inversify';
+import {Signal} from 'openfin-service-signal';
+
+import {APITopic, Events} from '../../client/internal';
+import {Inject} from '../common/Injectables';
+import {Injector} from '../common/Injector';
+
+import {APIHandler} from './APIHandler';
+import {CollectionMap} from './database/Database';
+
+export type AppInitData = {
+    id: string;
+    data: ApplicationOption | string
+}
+
+/**
+ * Client handler is responsible of keeping track of active clients, storing information about them
+ * and re-launching them when requested by other modules. Is intended to be used solely by EventPump.
+ *
+ */
+@injectable()
+export class ClientHandler {
+    private _apiHandler!: APIHandler<APITopic, Events>;
+
+    private _activeClients: Identity[] = [];
+    public readonly onClientHandshake: Signal<[Identity]> = new Signal();
+
+    constructor(@inject(Inject.API_HANDLER) apiHandler: APIHandler<APITopic, Events>) {
+        this._apiHandler = apiHandler;
+        this._apiHandler.onConnection.add(this.onClientConnect, this);
+        this._apiHandler.onDisconnection.add(this.onClientDisconnection, this);
+    }
+
+    /**
+     * Launches the app with specified identity if the start-up information about it was stored previously.
+     *
+     * @param app Identity of target client.
+     *
+     */
+    public tryLaunchApplication(app: Identity): void {
+        fin.Application.wrapSync(app).isRunning().then(running => {
+            if (!running) {
+                const database = Injector.get<'DATABASE'>(Inject.DATABASE);
+                database.get(CollectionMap.CLIENTS).get(app.uuid).then(value => {
+                    if (value) {
+                        (typeof value.data === 'string') ? fin.Application.startFromManifest(value.data) : fin.Application.start(value.data);
```

Comment:
Take care with exceptions here. If the app has a problem starting up, this could throw within the provider.

Reply:
done